board/safety/safety_toyota.h

```diff
@@ -29,14 +35,59 @@ const int32_t RT_INTERVAL = 250000;    // 250ms between real time checks
 const int16_t MAX_ACCEL = 1500;        // 1.5 m/s2
 const int16_t MIN_ACCEL = -3000;       // 3.0 m/s2
 
+int cruise_engaged_last = 0;           // cruise state
+int ipas_state = 1;                    // 1 disabled, 3 executing angle control, 5 override
+int angle_control = 0;                 // 1 if direct angle control packets are seen
+double speed = 0.;
+
+// track the torque measured for limiting
+int16_t torque_meas[3] = {0, 0, 0};    // last 3 motor torques produced by the eps
+int16_t torque_meas_min = 0, torque_meas_max = 0;
+int16_t angle_meas[3] = {0, 0, 0};     // last 3 steer angles
+int16_t angle_meas_min = 0, angle_meas_max = 0;
+int16_t torque_driver[3] = {0, 0, 0};    // last 3 driver steering torque
+int16_t torque_driver_min = 0, torque_driver_max = 0;
+
 // global actuation limit state
 int actuation_limits = 1;              // by default steer limits are imposed
 int16_t dbc_eps_torque_factor = 100;   // conversion factor for STEER_TORQUE_EPS in %: see dbc file
 
 // state of torque limits
 int16_t desired_torque_last = 0;       // last desired steer torque
+int16_t desired_angle_last = 0;        // last desired steer angle
 int16_t rt_torque_last = 0;            // last desired torque for real time check
+int16_t rt_angle_last = 0;             // last desired torque for real time check
 uint32_t ts_last = 0;
+uint32_t ts_angle_last = 0;
+
+int controls_allowed_last = 0;
+
+int to_signed(int d, int bits) {
+  if (d >= (1 << (bits - 1))) {
+    d -= (1 << bits);
+  }
+  return d;
+}
+
+// interp function that holds extreme values
+double interpolate(struct Lookup xy, double x) {
+  int size = sizeof(xy.x) / sizeof(xy.x[0]);
+  if (x <= xy.x[0]) {
+    return xy.y[0];
+
+  } else {
+    for (int i=0; i < size-1; i++) {
+      if (x < xy.x[i+1]) {
+        return (xy.y[i+1] - xy.y[i]) * (x - xy.x[i]) / (xy.x[i+1] - xy.x[i]) + xy.y[i];
```

Comment:
This is unreadable, and it's not immediately obvious why it won't divide by 0 (I don't think it will, but that should be so obvious)